overlord/hookstate/hooks.go

```diff
@@ -72,6 +78,189 @@ func SetupPreRefreshHook(st *state.State, snapName string) *state.Task {
 	return task
 }
 
+type gateAutoRefreshHookHandler struct {
+	context             *Context
+	refreshAppAwareness bool
+}
+
+func (h *gateAutoRefreshHookHandler) Before() error {
+	st := h.context.State()
+	st.Lock()
+	defer st.Unlock()
+
+	tr := config.NewTransaction(st)
+	experimentalRefreshAppAwareness, err := features.Flag(tr, features.RefreshAppAwareness)
+	if err != nil && !config.IsNoOption(err) {
+		return err
+	}
+	if !experimentalRefreshAppAwareness {
+		return nil
+	}
+
+	h.refreshAppAwareness = true
+
+	snapName := h.context.InstanceName()
+
+	// obtain snap lock before manipulating runinhibit lock.
+	lock, err := snaplock.OpenLock(snapName)
+	if err != nil {
+		return err
+	}
+	if err := lock.Lock(); err != nil {
+		return err
+	}
+	defer lock.Unlock()
+
+	if err := runinhibit.LockWithHint(snapName, runinhibit.HintInhibitedGateRefresh); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (h *gateAutoRefreshHookHandler) Done() (err error) {
+	ctx := h.context
+	st := ctx.State()
+	ctx.Lock()
+	defer ctx.Unlock()
+
+	snapName := h.context.InstanceName()
+
+	var action snapstate.GateAutoRefreshAction
+	a := ctx.Cached("action")
+
+	// obtain snap lock before manipulating runinhibit lock.
+	var lock *osutil.FileLock
+	if h.refreshAppAwareness {
+		lock, err = snaplock.OpenLock(snapName)
+		if err != nil {
+			return err
+		}
+		if err := lock.Lock(); err != nil {
+			return err
+		}
+		defer lock.Unlock()
+	}
+
+	// default behavior if action is not set
+	if a == nil {
+		// action is not set if the gate-auto-refresh hook exits 0 without
+		// invoking --hold/--proceed; this means proceed (except for respecting
+		// refresh inhibit).
+		if h.refreshAppAwareness {
+			if err := runinhibit.Unlock(snapName); err != nil {
+				return fmt.Errorf("cannot unlock inhibit lock for snap %s: %v", snapName, err)
+			}
+		}
+		return snapstate.ProceedWithRefresh(st, snapName)
+	} else {
+		var ok bool
+		action, ok = a.(snapstate.GateAutoRefreshAction)
+		if !ok {
+			return fmt.Errorf("internal error: unexpected action type %T", a)
+		}
+	}
+
+	// action is set if snapctl refresh --hold/--proceed was called from the hook.
+	switch action {
+	case snapstate.GateAutoRefreshHold:
+		// for action=hold the ctlcmd calls HoldRefresh; only unlock runinhibit.
+		if h.refreshAppAwareness {
+			if err := runinhibit.Unlock(snapName); err != nil {
+				return fmt.Errorf("cannot unlock inhibit lock of snap %s: %v", snapName, err)
+			}
+		}
+	case snapstate.GateAutoRefreshProceed:
+		// for action=proceed the ctlcmd doesn't call ProceedWithRefresh
+		// immediately, do it here.
+		if err := snapstate.ProceedWithRefresh(st, snapName); err != nil {
+			return err
+		}
+		if h.refreshAppAwareness {
+			// we have HintInhibitedGateRefresh lock already when running the hook,
+			// change it to HintInhibitedForRefresh.
+			if err := runinhibit.LockWithHint(snapName, runinhibit.HintInhibitedForRefresh); err != nil {
```

Comment:
this looks right, also not for this PR maybe but we need following the last discussion in the spec, make sure that the soft check and hard check just proceed with the lock is already from here with this value!

Reply:
I will address it in a followup. Thanks.